daemons/controld/controld_schedulerd.c

```diff
@@ -110,67 +111,94 @@ pe_ipc_destroy(gpointer user_data)
+         * fsa_input.msg that it is expecting. The scheduler's IPC dispatch function
+         * gave us the values we need, we just need to put them into XML.
+         *
+         * The name of the top level element here is irrelevant.  Nothing checks it.
+         */
+        fsa_input.msg = create_xml_node(NULL, "dummy-reply");
+        crm_xml_add(fsa_input.msg, XML_ATTR_REFERENCE, msg_ref);
+        crm_xml_add(fsa_input.msg, F_CRM_TGRAPH_INPUT, reply->data.graph.input);
+
+        crm_data_node = create_xml_node(fsa_input.msg, F_CRM_DATA);
+        add_node_nocopy(crm_data_node, NULL, reply->data.graph.tgraph);
+        register_fsa_input_later(C_IPC_MESSAGE, I_PE_SUCCESS, &fsa_input);
+
+    } else {
+        crm_info("%s calculation %s is obsolete", CRM_OP_PECALC, msg_ref);
+    }
+}
+
+static void
+scheduler_event_callback(pcmk_ipc_api_t *api, enum pcmk_ipc_event event_type,
+                         crm_exit_t status, void *event_data, void *user_data)
+{
+    pcmk_schedulerd_api_reply_t *reply = event_data;
+
+    switch (event_type) {
+        case pcmk_ipc_event_disconnect:
+            pe_ipc_destroy();
+            break;
+
+        case pcmk_ipc_event_reply:
+            handle_reply(reply);
+            break;
+
+        default:
+            break;
     }
-    free_xml(msg);
-    return 0;
 }
 
-/*!
- * \internal
- * \brief Make new connection to scheduler
- *
- * \return TRUE on success, FALSE otherwise
- */
 static bool
 pe_subsystem_new(void)
 {
-    struct ipc_client_callbacks pe_callbacks = {
-        .dispatch = pe_ipc_dispatch,
-        .destroy = pe_ipc_destroy
-    };
-    static bool retry_one = TRUE;
+    int rc;
 
     controld_set_fsa_input_flags(R_PE_REQUIRED);
-retry:
-    pe_subsystem = mainloop_add_ipc_client(CRM_SYSTEM_PENGINE,
-                                           G_PRIORITY_DEFAULT,
-                                           5 * 1024 * 1024 /* 5MB */,
-                                           NULL, &pe_callbacks);
-    if (pe_subsystem == NULL) {
-        crm_debug("Could not connect to scheduler : %s(%d)", pcmk_rc_str(errno), errno);
-        if (errno == EAGAIN && retry_one) {
-            /* In rare cases, a SIGTERM may be received and the connection may fail when the cluster shuts down. */
-            /* At this time, the connection will be retried only once. */
-            crm_debug("Scheduler connection attempt.");
-            retry_one = FALSE;
-            goto retry;
-        }
-        return FALSE;
+    rc = pcmk_new_ipc_api(&schedulerd_api, pcmk_ipc_schedulerd);
+
+    if (rc != pcmk_rc_ok) {
+        crm_err("Error connecting to the scheduler: %s", pcmk_rc_str(rc));
+        return false;
+    }
+
+    pcmk_register_ipc_callback(schedulerd_api, scheduler_event_callback, NULL);
+    rc = pcmk_connect_ipc(schedulerd_api, pcmk_ipc_dispatch_main);
+    if (rc != pcmk_rc_ok) {
+        crm_err("Error connecting to the scheduler: %s", pcmk_rc_str(rc));
+        return false;
```

Comment:
I'm thinking we should free and null schedulerd_api here

Reply:
Fixed.